Read request ID from header and use in logs

Choraden opened the pull request:
The proxy is capable to associate the request id send in request header with logs.

mmatczuk commented on internal/martian/handler.go:
```diff
@@ -89,7 +89,11 @@ func (p proxyHandler) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
 	ctx := withSession(session)
 
-	outreq := req.Clone(ctx.addToContext(req.Context()))
+	cctx := ctx.addToContext(req.Context())
```
How about `outreq` modification if needed?

Choraden replied:
I did not want to make another copy of a request, but I can change it.

mmatczuk replied:
Maybe refactor context creation to a function?

mmatczuk commented on internal/martian/proxy.go:
```diff
@@ -368,7 +368,11 @@ func (p *Proxy) readRequest(ctx *Context, conn net.Conn, brw *bufio.ReadWriter)
 		}
 
-		req = req.WithContext(ctx.addToContext(req.Context()))
+		cctx := ctx.addToContext(req.Context())
```
Same here.

mmatczuk commented on internal/martian/header/reqid_modifier.go:
```diff
@@ -0,0 +1,21 @@
+)
+
+// NewRequestIDModifier adds X-Request-ID header to the request if it is not already present.
+func NewRequestIDModifier() martian.RequestModifierFunc {
```
👍

mmatczuk commented on internal/martian/httpspec/httpspec.go:
```diff
@@ -33,6 +33,7 @@ func NewStack(via string) (outer, inner *fifo.Group) {
 	outer.AddRequestModifier(hbhm)
 	outer.AddRequestModifier(header.NewForwardedModifier())
 	outer.AddRequestModifier(header.NewBadFramingModifier())
+	outer.AddRequestModifier(header.NewRequestIDModifier())
```
This is not part of `httpspec`!

Actually this should be controlled by a flag and maybe users should have control over name because adding X-Request-ID may cause some unexpected behavior.

Choraden replied:
You're right, it's not part of the spec. It was comfortable to put it here.

Choraden replied:
I will add a flag.

mmatczuk commented on internal/martian/proxy.go:
```diff
@@ -822,7 +822,20 @@ func (p *Proxy) connectHTTP(req *http.Request, proxyURL *url.URL) (res *http.Res
 		d = dialvia.HTTPProxy(p.dial, proxyURL)
 	}
-	d.ConnectRequestModifier = p.ConnectRequestModifier
+	d.ConnectRequestModifier = func(r *http.Request) error {
```
This does not belong here.

Choraden replied:
I know, but you need to somehow get the trace ID and CONNECT request won't have it in its context. That's why it's here.

mmatczuk replied:
I think it does not belong to Marian but Forwarder and should be implemented there.
There should be zero changes to martian code.

Choraden commented:
I added `add-request-id` flag that controls whether to pass trace id as X-Request-ID. 

Unfortunately, we can't reuse X-Request-ID from Forwarder, as it requires modifying request context - we would need to overwrite a pointer. 

I also dropped support for passing X-Request-ID in CONNECT requests, as it is infeasible outside `martian.Proxy`.

mmatczuk commented on internal/martian/proxy.go:
```diff
@@ -836,7 +850,19 @@ func (p *Proxy) connectHTTP(req *http.Request, proxyURL *url.URL) (res *http.Res
 		d = dialvia.HTTPProxy(p.dial, proxyURL)
 	}
-	d.ConnectRequestModifier = p.ConnectRequestModifier
+	d.ConnectRequestModifier = func(r *http.Request) error {
```
Maybe we should have it as a function `p.connectRequestModifier` WDYT?

Choraden replied:
The only problem is that we need to pass a context from the original request,  so it would be 
```go
p.connectRequestModifier(ctx context.Context)
```
but sure I can do that.

mmatczuk commented on bind/flag.go:
```diff
@@ -127,6 +127,15 @@ func HTTPProxyConfig(fs *pflag.FlagSet, cfg *forwarder.HTTPProxyConfig, lcfg *lo
+
+	fs.StringVar(&cfg.AddRequestIDHeader, "add-request-id-header", cfg.AddRequestIDHeader,
+		"<header>"+
+			"header name that the proxy will add to outgoing requests with id value. ")
```
```suggestion
The header name that the proxy will add to outgoing requests with internal request ID value.
```

Choraden commented on internal/martian/proxy.go:
```diff
@@ -104,6 +104,11 @@ type Proxy struct {
 	// CloseAfterReply closes the connection after the response has been sent.
 	CloseAfterReply bool
 
+	// ReuseRequestIDFromHeader specifies a special header name that the proxy will use to identify requests.
```
ok

mmatczuk commented:
I have two thoughts strange naming and overall structure.

I'd probably want to see: 

`--log-http-request-id-header` with default value of "x-request-id" which means if request has this ID use it in logs - this is a logging feature

`--proxy-header x-request-id: {{req.ID()}}` - this is a syntax that treats header values like text template that can be evaluated against the request. ID then in a special function to get the request ID. This is much more flexible and elegant.

Choraden commented:
This PR will add only the feature of reading request ID and using it in logs.

mmatczuk commented on internal/martian/proxy.go:
```diff
@@ -62,6 +62,11 @@ type Proxy struct {
+	// LogHTTPRequestIDHeader specifies a special header name that the proxy will use to identify requests.
+	// If the header is present in the request, the proxy will associate the value with the request in the logs.
+	// If empty, no action is taken, and the proxy will generate a new request ID.
+	LogHTTPRequestIDHeader string
```
I think here it should be `RequestIDHeader`.

mmatczuk commented on http_proxy.go:
```diff
@@ -86,6 +86,7 @@ type HTTPProxyConfig struct {
 	CloseAfterReply        bool
 	DenyDomains            []*regexp.Regexp
 	DenyDomainsExclude     []*regexp.Regexp
+	LogHTTPRequestIDHeader string
```
I do not like the practice of adding to the end of the list in random moments it does not make sense.

I think here it should also be named RequestIDHeader and live above RequestModifiers. WDYT?

Choraden replied:
Will do it in #414

mmatczuk commented:
[http_proxy: enable LogHTTPRequestIDHeader](https://github.com/saucelabs/forwarder/pull/407/commits/7107591116ffec179731b5a60664b3937afa8dd3) and the last commit can be squashed.
